src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp

```diff
@@ -931,280 +1019,393 @@ void CompositionalMultiphaseWell::initializeWells( DomainPartition & domain )
-      arrayView1d< real64 > const & connRate = subRegion.getField< fields::well::mixtureConnectionRate >();
-
-      // get the info stored on well elements
-      arrayView2d< real64, compflow::USD_COMP > const & wellElemCompFrac = subRegion.getField< fields::well::globalCompFraction >();
-      arrayView1d< real64 const > const & wellElemGravCoef = subRegion.getField< fields::well::gravityCoefficient >();
-
-      // get the element region, subregion, index
-      arrayView1d< localIndex const > const resElementRegion = perforationData.getField< fields::perforation::reservoirElementRegion >();
-      arrayView1d< localIndex const > const resElementSubRegion = perforationData.getField< fields::perforation::reservoirElementSubRegion >();
-      arrayView1d< localIndex const > const resElementIndex = perforationData.getField< fields::perforation::reservoirElementIndex >();
-
-      arrayView1d< real64 const > const & perfGravCoef = perforationData.getField< fields::well::gravityCoefficient >();
-
-      // 1) Loop over all perforations to compute an average mixture density and component fraction
-      // 2) Initialize the reference pressure
-      // 3) Estimate the pressures in the well elements using the average density
-      PresTempCompFracInitializationKernel::
-        launch( perforationData.size(),
-                subRegion.size(),
-                numComp,
-                numPhase,
-                perforationData.getNumPerforationsGlobal(),
-                wellControls,
-                0.0, // initialization done at t = 0
-                resCompFlowAccessors.get( fields::flow::pressure{} ),
-                resCompFlowAccessors.get( fields::flow::temperature{} ),
-                resCompFlowAccessors.get( fields::flow::globalCompDensity{} ),
-                resCompFlowAccessors.get( fields::flow::phaseVolumeFraction{} ),
-                resMultiFluidAccessors.get( fields::multifluid::phaseMassDensity{} ),
-                resElementRegion,
-                resElementSubRegion,
-                resElementIndex,
-                perfGravCoef,
-                wellElemGravCoef,
-                wellElemPressure,
-                wellElemTemp,
-                wellElemCompFrac );
-
-      // get well secondary variables on well elements
-      string const & fluidName = subRegion.getReference< string >( viewKeyStruct::fluidNamesString() );
-      MultiFluidBase & fluid = getConstitutiveModel< MultiFluidBase >( subRegion, fluidName );
-      arrayView3d< real64 const, multifluid::USD_PHASE > const & wellElemPhaseDens = fluid.phaseDensity();
-      arrayView2d< real64 const, multifluid::USD_FLUID > const & wellElemTotalDens = fluid.totalDensity();
 
-      // 4) Back calculate component densities
-      constitutive::constitutiveUpdatePassThru( fluid, [&] ( auto & castedFluid )
+      if( time_n <= 0.0 ||
```

Comment:
why?
what about wells that open later for time>0?

Reply:
full check is      if( time_n <= 0.0 ||
          ( !wellControls.isWellOpen( time_n ) && wellControls.isWellOpen( time_n + dt ) )  ) Previously a well was only initialized at time <=0.   the or part allows initialization for a well that is currently closed and next dt its scheduled to be opened.  Models containing wells using rate tables with first nonzero rate at t>0 were failing to converge.

Reply:
i see, yeah, that part is tricky...